### PR TITLE
Text scoring engine

### PR DESCRIPTION
**Pull request by Sh3llSp4wn**
Added Text Scoring Engine to `Ronin::Support::Text::Scoring`

### CODE REVIEW

**Comment by postmodern** [comment]
Some questions about this:

1. What does this Text Scoring Engine measure exactly? Text Scoring Engine is a little vague. What specific algorithm is it implementing?
2. How would this Text Scoring Engine be used? When would someone need to use this Text Scoring Engine? What problem(s) does this Text Scoring Engine solve?
3. Does this really need to be in `ronin-support`? Is it related to security? It could be published as it's own gem which you would fully own and control.

**Comment by Sh3llSp4wn** [comment]
The text scoring engine is a thin helper class for searching possible "answer spaces" programmatically. 

I've used it in a lot of CTFs that feature or include cryptography as a more general solution to searching for English text.

The idea is that you provide a `metric` and a `dataset`. The metric is run against each element of the dataset. Once recorded the scores are then sorted for access to best match, worst match, and/or top x matches. 

All this is wrapped in a helper function I hope will be more directly useful. In `Ronin::Support::Text::Scoring` there is a function called `find_english_text` which takes an array of strings and returns the one that is most english like. 

I built the system around these swappable metrics because I wanted to be able to come in later with spanish, french, etc datasets. It is also a little more general than that, because the Score Engine doesn't actually care what the thing it is scoring is. 

Security relevancy is in similar keeping with the other items in `text.rb` and `text` which provide exploratory tools for dealing with unknown data.  

An example of this code is a solution to Set 1 Challenge 4 of crypto_pals (I generated the inital script with `ronin new`)

```ruby
require 'ronin/support'
include Ronin::Support

require 'awesome_print'

best = ['', 0]

# grab each line from 4.txt
File.readlines('4.txt').map do |line|
  # extract best english canidate from each line after xor decryption
  new_canidate = Ronin::Support::Text::Scoring.find_english_text(
    (0..255).map { |k| line.hex_decode.chars.map { |c| c ^ k }.join }
  )
  # simple who is best swap
  best = new_canidate if new_canidate[1] > best[1]
  print '.' # status
end

ap best
```

**Comment by postmodern** [comment]
@Sh3llSp4wn hmm, I am a bit nervous about merging a custom algorithm, which would make me the maintainer of it. We already provide a `Text::Entropy` module and methods for filtering de-ciphered text based on entropy, which *should* have the same effect as filtering by textual language. I would instead encourage you to turn this into your own gem, ideally with a more descriptive name like `language-scoring`. That way you retain full ownership of the copyright.

But I will provide feedback on the code before closing this PR to help you out.

File: lib/ronin/support/text/text_score.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+module Ronin
+  module Support
+    module Text
+      #
+      # Scoring provides basic text scoring capibilites for
+      # english. Can be extended to other languages or formats
+      # using provided generator script in
+      # `ronin/support/text/text_score/`
+      #
```
Comment (postmodern) [inline lib/ronin/support/text/text_score.rb:13]: Consider adding an explanation of how text is parsed and then scored.

File: lib/ronin/support/text/text_score.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+require 'ronin/support/text/text_score/engine'
+
+module Ronin
+  module Support
+    module Text
+      #
+      # Scoring provides basic text scoring capibilites for
+      # english. Can be extended to other languages or formats
```
Comment (postmodern) [inline lib/ronin/support/text/text_score.rb:10]: Consider making it language agnostic or support a `lang:` keyword argument.

File: lib/ronin/support/text/text_score.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      # using provided generator script in
+      # `ronin/support/text/text_score/`
+      #
+      # You'll need to modifiy the erb template to specify language
+      # if you plan to do this later. @todo: shellspawn
+      #
+      module Scoring
+        #
```
Comment (postmodern) [inline lib/ronin/support/text/text_score.rb:18]: Missing a summary of what the method does.

File: lib/ronin/support/text/text_score.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        # @param [Array<String>] data_set
+        #   the array of english text canidates
+        #
+        # @return [<String>, Score] or Array[[<String>, Score]]
+        #   Score is the result of the provided metric applied to
+        #   the string. This function returns the one that looks
+        #   most like english, or the top_n number of scores
+        #
+        def self.find_english_text(data_set, top_n=1)
```
Comment (postmodern) [inline lib/ronin/support/text/text_score.rb:29]: Consider using a keyword argument for `top_n:` to make it clear what the passed in argument is for.

File: lib/ronin/support/text/text_score.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+          # this could be made a bit nicer by passing a symbol to the
+          # constructor, but this suffices TODO
+          tse = TextScoringEngine.new { |e| english_n_glyph_metric e }
```
Comment (postmodern) [inline lib/ronin/support/text/text_score.rb:32]: I encourage using parenthesis for method calls, unless they are declarative methods such as `attr_reader` or `system`.

File: lib/ronin/support/text/text_score/engine.rb [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+      # It then selects the elements of the dataset that best match the metric.
+      #
+      # Selected elements are avalible via `best`, `worst`. `top`, and etc
+      #
+      #
+      # @api public
+      module Scoring
+        # a nice default metric for use with TSE
+        #
+        # @parameter element
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:21]: `@param` not `@parameter`. Also needs a type signature for `element`.

File: lib/ronin/support/text/text_score/engine.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        #   this opaque element param expects a string,
+        #   which is provided by the score function of TSE
+        # @return float
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:24]: Use capitalize `Float`.

File: lib/ronin/support/text/text_score/generate_glyphs.rb [hunk 8]
```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+require 'awesome_print'
+require 'erb'
+
+# Computes n-glyphs from a corpus of text
+class NGlyphFrequencyAnalyzer
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/generate_glyphs.rb:6]: I would strongly discourage putting internal utility code into `lib/` as any other gem can require that code. Instead, you should define this as a `rake` task to generate the files, ideally using Rake's `file` task.

Also, it's not a good idea to define classes at the top-level within the `lib/` directory.

File: lib/ronin/support/text/text_score/generate_glyphs.rb [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    # NOTICE: this script is generated by generate_glyphs.rb #
+    ##########################################################
+
+    module Ronin
+      module Support
+        module Text
+          # Negitive constant to make random data look even less like english
+          # -phi because I was just working on fib calculators
+          NOISE_CONSTANT = -1.618
+        <% analyzers.each do | analyzer | %>
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/generate_glyphs.rb:93]: You might be able to lose the dependency on `erb` (although it is included in the stdlib), by rewriting the code to append lines to a `lines = []` buffer and join it using `.join($/)`.

File: spec/text/text_score/engine_spec.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
+require 'spec_helper'
+require 'ronin/support/encoding/base64'
+
+describe Ronin::Support::Text::Scoring do
+  describe ".english_n_glyph_metric" do
+    context "when providing english text" do
+      it "must return a positive numeric score" do
+        expect(subject.english_n_glyph_metric("eeeeeeeeeee")).to be_a_kind_of(Numeric)
+        expect(subject.english_n_glyph_metric("eeeeeeeeeee")).to be > 0
+      end
```
Comment (postmodern) [inline spec/text/text_score/engine_spec.rb:10]: I would add more tests for positive-matching edge-cases such as:

* a single word
* two words
* a sentence
* a sentence with a comma
* a sentence with a semi-colon.
* two sentences.
* two sentences joined by a newline.

File: lib/ronin/support/text/text_score/engine.rb [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+          # Initialize an new scoring engine
+          #
+          # NOTE: the only requirement for
+          # the opaque elements used in this
+          # engine is hashability
+          #
+          # @param [Proc<Element>] metric
+          #   The metric, as a block or Proc
+          #   that scores the elements
+          def initialize(&metric)
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:56]: Instead of passing in a `Proc`, what about passing in some kind of `Language` module which contains all of the glyph statistics as well as the metric measuring method?

File: lib/ronin/support/text/text_score/engine.rb [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+          end
+        end
+
+        #
+        # Text Scoring engine implementation
+        class TextScoringEngine
+          #
+          # Initialize an new scoring engine
+          #
+          # NOTE: the only requirement for
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:49]: Use a YARD `@note` tag for notes.

File: lib/ronin/support/text/text_score/engine.rb [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+            raise ArgumentError, "Must provide metric as a block or Proc" if metric.nil?
+
+            @metric          = metric
+            @dataset         = []
+            @scores          = Hash.new(0)
+          end
+
+          #
+          # add an element to process
+
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:66]: Random empty line.

File: lib/ronin/support/text/text_score/engine.rb [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+          # @param <Element> elements
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:67]: Use a consistent YARD style of `@param [Type] value` instead of mixing different styles.

File: lib/ronin/support/text/text_score/engine.rb [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+      #
+      # The text scoring engine takes a "metric" and a "dataset" as parameters.
+      # It then selects the elements of the dataset that best match the metric.
+      #
+      # Selected elements are avalible via `best`, `worst`. `top`, and etc
+      #
+      #
+      # @api public
+      module Scoring
+        # a nice default metric for use with TSE
```
Comment (postmodern) [inline lib/ronin/support/text/text_score/engine.rb:19]: Would be nice if there was more of a description on what this method does.

**Review by postmodern — COMMENTED** [review]
Submitted some feedback to help you refine this code further and hopefully release it as it's own gem.
